### ISSUE TITLE
does forward/eval from a trained mamba model require cuda as well?

### ISSUE THREAD

**Issue by shadowleaves** [state-spaces/mamba#89]
codes in selective_scan_fwd() of selective_scan.cpp seem to suggest even forward from a trained model would require cuda, which might be inconvenient when running models in production environments. Any idea how to do model forward on a CPU-only machine? Thanks

**Comment by tridao** [comment]
Yup, it's only implemented for CUDA for now. You can look at the `selective_scan_ref` for the pure pytorch implementation that should run on CPU (though probably quite slow).

**Comment by shadowleaves** [comment]
thanks, will look into it

**Comment by kroggen** [comment]
You can check [this fork](https://github.com/kroggen/mamba-cpu). It works on CPU

**Comment by kroggen** [comment]
Inference of Mamba models in pure C

https://github.com/kroggen/mamba.c

Recurrent mode only, for simplicity

Faster than pytorch (in default mode) on CPU
